### ISSUE TITLE
Maintaining Cross Domain session with passport.js

### ISSUE THREAD

**Issue by Prateek479** [jaredhanson/passport#403]
I am using nodeJs for server and with express and passport as authentication middleware. My front end app is running on angular on localhost:9000 and node app on localhost:3000 . I found that the passport is not able to save session in cross domain . on every page request on my client app , passport create a new session for me instead of maintaining a single session . Also after login on login page it create a session in mongo with user detail available in it , but on second request it never resolve the same session cookie for me  .

**Comment by caal-15** [comment]
I'm having the same issue, did you find a solution?, if not I would really appreciate any help, the login works but the session is not maintained, I have already made sure all headers are fine and that the cookie actually gets to the backend, but it still doesn't work

**Comment by AntonScotte** [comment]
Have you been able to find a solution for this? I'm facing the same problem.

**Comment by apmeena** [comment]
+1

**Comment by apmeena** [comment]
I have tried 

```
var cors = require('cors');
var app = express();
app.use(cors());
```

and 

```
app.all('*', function(req, res, next) {
       res.header("Access-Control-Allow-Origin", "*");
       res.header("Access-Control-Allow-Headers", "X-Requested-With");
       res.header('Access-Control-Allow-Headers', 'Content-Type');
       next();
});
```

But neither of them worked.

**Comment by jmls** [comment]
did anyone solve this ?

**Comment by kingdevnl** [comment]
+1

**Comment by JoBoDo** [comment]
Also +1

**Comment by nitzano** [comment]
You'll need SSO (single sign on) implementation for that using tokens exchange,
 however if you use subdomains and not two different domains it is possible to get a shared session by setting up a global session store (e.g with redis) and setting a cookie to the root domain like this :

```javascript
const expressSession = require('express-session');
const redisStore = require('connect-redis')(expressSession);
...
app.use(
  expressSession({
    store: new redisStore({ host: 'localhost', port: 6379 }),
    secret: 'keyboard cat',
    cookie: { domain: 'domain.com' },
    resave: false,
    saveUninitialized: false
  })
);
app.use(passport.initialize());
app.use(passport.session());
```

This way you'll get shared session for both `app1.domain.com` and `app2.domain.com`

However for cross domains you'll need SSO implementation with another authorization server (e.g `sso.otherdomain.com`) which will exchange tokens for each domain after authenticating to to the authentication server. I'm trying to find an example/documentation for passport with SSO for myself as well but couldn't find anything so far (closest I've found is [this](https://codeburst.io/building-a-simple-single-sign-on-sso-server-and-solution-from-scratch-in-node-js-ea6ee5fdf340))

**Comment by YUzhva** [comment]
For making cross-domain requests we are using the proxy, to avoid of CORS issue.

TL;DR
**I think that issue is in the proxy. And we need to set up some "reverse proxy" on NodeJS API side.**

I'm also using separate apps:
For server API: http:// localhost:3000/
For client SPA: http://localhost:8080/

When I'm sending request from http://localhost:8080/ I'm using `/api-local/` prefix for `http-proxy-middleware`, so my config looks like:
```
'/api-local/': {
  target: http:// localhost:3000/,
  pathRewrite: { '^/api-local/': '/' },
  changeOrigin: true,
}
```

So the next request in SPA `fetch('/auth')` will be interpreted as `fech('http:// localhost:3000/auth')`.

The trick is that authentication at NodeJS side works very well, all data from LinkedIn are returns and the *passportjs* is invoke `serializeuser` method.

When I'm sending the second request after authentication (it does not matter if it is protected or unprotected API request) - API after receiving a request, invoking `deserializeUser` passportjs method.

Only when I'm disabling passport session with `{ session: false }` option - it works ok.

So there are few conclusions:
- passportjs make its job with authentication
- Next request after authentication works only with user wast authenticated without a session
- there is something wrong with the session *OR during next requests, information about user session is missing in request*
